### PR TITLE
Roots finding

### PR DESCRIPTION
**Pull request by DartVart**

### CODE REVIEW

File: roots_finding.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            segments.append(LineSegment(cur_segment.left, cur_segment.right))
+        left_function_value = right_function_value
+        cur_segment.left = cur_segment.right
+
+    return segments
+
+
+class SingleRootSolver(ABC):
+    @abstractmethod
+    def find_root(self, function, line_segment: LineSegment, accuracy, variable: str = 'x') -> float:
```
Comment (GirZ0n) [inline roots_finding.py]: Please add typing for `accuracy`.

File: roots_finding.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        pass
+
+
+# It is assumed that there is exactly one non-multiple root inside the line segment
+class HalfDivisionSolver(SingleRootSolver):
+    def __init__(self):
+        self.number_of_iterations = 0
+
+    def find_root(self, function, line_segment: LineSegment, accuracy, variable: str = 'x') -> float:
```
Comment (GirZ0n) [inline roots_finding.py]: Please add typing for `accuracy`.

File: roots_finding.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            center_function_value = function.subs({variable: cur_segment.center})
+
+            if left_function_value * center_function_value <= 0:
+                cur_segment.right = cur_segment.center
+                right_function_value = center_function_value
+            elif center_function_value * right_function_value <= 0:
+                cur_segment.left = cur_segment.center
+                left_function_value = center_function_value
+            else:
+                raise ValueError("Some error")
```
Comment (GirZ0n) [inline roots_finding.py]: Please use a clearer error description

File: roots_finding.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        return cur_segment.center
+
+
+class Solver:
+    def __init__(self, single_root_solver):
+        self.single_root_solver = single_root_solver
+
+    def find_roots(self, function, line_segment, accuracy, number_of_steps, variable: str = 'x'):
```
Comment (GirZ0n) [inline roots_finding.py]: Please add typing

File: roots_finding.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                cur_segment.left = cur_segment.center
+                left_function_value = center_function_value
+            else:
+                raise ValueError("Some error")
+
+        return cur_segment.center
+
+
+class Solver:
+    def __init__(self, single_root_solver):
```
Comment (GirZ0n) [inline roots_finding.py]: Please add typing

File: roots_finding.py (hunk 1 above)
Comment (DartVart) [reply roots_finding.py]: No

File: roots_finding.py (hunk 2 above)
Comment (DartVart) [reply roots_finding.py]: No No

File: roots_finding.py (hunk 3 above)
Comment (DartVart) [reply roots_finding.py]: No

File: roots_finding.py (hunk 1 above)
Comment (GirZ0n) [reply roots_finding.py]: 🗿

File: roots_finding.py (hunk 2 above)
Comment (GirZ0n) [reply roots_finding.py]: 🗿

File: roots_finding.py (hunk 3 above)
Comment (GirZ0n) [reply roots_finding.py]: 🗿

File: roots_finding.py (hunk 1 above)
Comment (DartVart) [reply roots_finding.py]: Okay, I thought and changed my mind. Will be done!

File: roots_finding.py (hunk 2 above)
Comment (DartVart) [reply roots_finding.py]: Okay, I thought and changed my mind. Will be done!

File: roots_finding.py (hunk 3 above)
Comment (DartVart) [reply roots_finding.py]: Okay, I thought and changed my mind. Will be done!

File: roots_finding.py (hunk 4 above)
Comment (DartVart) [reply roots_finding.py]: Ok

File: roots_finding.py (hunk 5 above)
Comment (DartVart) [reply roots_finding.py]: Ok

**Comment by illided** [comment]
I can't see any tests and linkers. Please add enterprise continuous integration with linker, tests, performance benchmarks and auto updating documentation.

**Comment by DartVart** [comment]
cat

**Comment by Furetur** [comment]
It is almost impossible to use this on my machine. Please provide a docker image with all the necessary instructions.

Also I suggest adding some enterprise loggers and building everything with Gradle

**Comment by DartVart** [comment]
pocc

**Comment by GirZ0n** [comment]
image

**Pull request merged**